### ISSUE TITLE
Implement an ORM

### ISSUE THREAD

**Issue by prhost** [copona/copona#31]
I think it is essential and necessary to have an ORM to make it easier for both our lives and those of the developers. Creating pure query is already outdated and there is nothing reusable. It also does not help in modularization.

I have always worked with Eloquent ORM that is of laravel and uses Doctrine in its core, it is an independent library easy to install. We can even create a driver for it at first and keep the models as they are, and go transcribing gradually.

Https://github.com/illuminate/database

There are other ORMs that are easy to handle as propel.

Doctrine always thought this ORM was made by NASA. Very complex.

**Comment by arnisjuraga** [comment]
I agree on DB driver improvements. But there is a catch. ORM is good, elegant and very, very powerful. Although - not an easy to learn and requires experience not only with OOP, DB and PHP, but also with a a specific ORM library itself.

The biggest strength of Opencart is it's "simplicity". MVC architecture is simple and straight forward. Even developer beginner can take it as source and adapt it quite fast for his needs without deep knowledge of OOP or OC documentation.

This is the same issue as for Composer - we could implement it, but it makes project much more complex, not affordable for many "standard" user. E.g. - shared hostings usually has no SSH support and so on.

There are other E-commerces, which are much more advanced at this point and I doubt that Copona would like to rush into that field. 

I would suggest to start with implementation of PDO library. 
Opencart is "kind a" based on Codeigniter. It does not matter - is it MySQLi PDO, Codeigniter ActiveRecord or even Doctrine support. But completely rewriting SQL models for support of ORM wouldn't be in a scope of Copona. 

What do @copona/ops  think?

**Comment by prhost** [comment]
It sounds like Daniel's vision, thinking that composer is something from the other world, that's what made him stand still in the past. Composer has been around for a long time, and it's the least a project should have. I think that thinking about apprentices, who program structured and do not know how to evolve in the area is not a good way, that stains the face of the platform, that's what happened to Opencart, nowadays it's not so famous because it does not evolve, it's no use I think there is a lot more people wanting to grow and always improve their techniques, that's why php invented the PSR, to create patterns, get out of the clutter and be seen again as a language Robust About shared hosting is the same thing, I know there are many, but not something good, something that will be the future of hosting, and outside that does not need to run composer in the hosting environment, it is not even a good practice of Devops, composer Must be run before along with any build and only sent to the hosting the project artifact. No unnecessary files like composer.json, gitignore, etc.

On the ORM that has eluded the subject a little, the idea is to make a driver, it will work exactly like ta, with the huge and ugly query that has in the models. The only difference is that the developer will also be able to have ORM support for their models along with a complete documentation that is the laravel is own website for example if he wants to use it uses if he can not use the good old `$ this->db->query()` That wanting or not any ORM supports raw query.

**Comment by arnisjuraga** [comment]
You got me with  `It sounds like Daniel's vision` ha ha ha :) 

Ok, to be more clear - I am not against and you are completely right about `because it does not evolve`. I'm just cautious on not to rush things more then we can handle :)

Anyway - Composer definitely is something we must have. But ORM is something, we _should_ have along more simple way of managing DB. Because it will take huge amount of work to do it and must be a plan for the further.

**Comment by prhost** [comment]
I understand your concern hehe, it will be difficult to hold me in relation to implementing new things. I love to innovate and know what's new in the PHP world. So hold on hehe. But I do understand what you are worried about, that is why I said that I can keep both, preserve the old one and migrate gradually until one day we decide to leave at once. Believe me, ORM mainly Eloquent is very good, fast to develop and will attract good developers, we can take advantage of the event functionality that has internally in it to modularize and provide even more as the developer inject code without needing the damn VQMOD / OCMOD (:

**Comment by arnisjuraga** [comment]
ok, @prhost - let's create a "project" for ORM iplementation, and see how it's going.

**Comment by arnisjuraga** [comment]
@prhost 2y old project, but still. https://github.com/oney/opencart-laravel
